File: src/respiration.js
'use strict';

// Respiration-rate estimation from RR intervals via Respiratory Sinus
// Arrhythmia (RSA): breathing modulates heart rate (HR rises on inhale, falls
// on exhale), so the RR-interval series oscillates at the breathing frequency.
//
// Pipeline (Welch PSD):
//   tachogram (beat time, RR) -> linear-interpolate to 4 Hz -> 2nd-order
//   polynomial detrend (kills DC + slow non-linear drift that leaks into the
//   respiration band) -> split into 50%-overlapping Hann-windowed segments ->
//   average the per-segment periodograms (Welch) -> scan the search band for
//   the dominant peak.
//
// Confidence is a SIGNAL-QUALITY score, not a probability: it combines the
// peak SNR (peak power / median band power = robust noise floor) and the peak
// sharpness (narrow peak = a real oscillation, broad = noise). Welch averaging
// reduces the periodogram variance so weak RSA (low-RMSSD subjects) still
// produces a stable, separable peak — a single periodogram is too noisy there.
//
// Bands are deliberately split:
//   - SEARCH band 0.10-0.50 Hz (6-30 br/min) so we can also catch slow/fast
//     breathing, not just the classic HF band.
//   - HRV HF band 0.15-0.40 Hz is the standard vagal HF interpretation band
//     (Task Force 1996); peaks below ~0.13 Hz risk Mayer-wave (LF, ~0.1 Hz)
//     confusion and are de-weighted.
//
// Refs: Welch (1967); Task Force HRV standards (1996); Schäfer & Kratky (2008)
// "Estimation of breathing rate from RSA"; Natarajan et al. (2021) npj Digital
// Med (PSD-background + SNR gating for wearable respiration).

const DEFAULTS = {
  fs: 4, // Hz, resample grid
  // Search band starts at the classic HF lower bound 0.15 Hz (9 br/min): going
  // lower invites the Mayer wave (~0.1 Hz baroreflex oscillation), which is a
  // strong, clean LF peak that is NOT respiration — at rest/coding it would be
  // mistaken for "6 br/min". Slow paced breathing < 9/min is out of scope here.
  searchMin: 0.15, // Hz, peak-search lower bound (9 br/min)
  searchMax: 0.45, // Hz, peak-search upper bound (27 br/min, covers fast/shallow)
  hfMin: 0.15, // Hz, HRV HF band (interpretation only)
  hfMax: 0.4,
  mayerMax: 0.16, // Hz, near the LF/HF boundary -> de-weight a weak edge peak
  step: 0.005, // Hz, scan resolution (oversamples the true bin width)
  segSec: 60, // s, Welch segment length
  overlap: 0.5, // Welch segment overlap fraction
  previewSpanMs: 30000, // 30-60 s -> preview (provisional, low confidence)
  minSpanMs: 60000, // >= 60 s -> a full estimate (>=4 breaths even at 6/min)
  minEntries: 20, // and enough beats
  minSnr: 2.5, // peak must exceed the median noise floor by this factor
};

function resampleLinear(xs, ys, fs, t0, tEnd) {
  const dtMs = 1000 / fs;
  const n = Math.max(2, Math.floor((tEnd - t0) / dtMs) + 1);
  const out = new Float64Array(n);
  let j = 0;
  for (let i = 0; i < n; i++) {
    const t = t0 + i * dtMs;
    while (j < xs.length - 2 && xs[j + 1] < t) j++;
    const x0 = xs[j];
    const x1 = xs[j + 1];
    const frac = x1 > x0 ? (t - x0) / (x1 - x0) : 0;
    out[i] = ys[j] + (ys[j + 1] - ys[j]) * Math.max(0, Math.min(1, frac));
  }
  return out;
}

// 2nd-order polynomial detrend: removes DC, linear and quadratic trend so a
// slowly drifting mean HR does not leak power into the respiration band.
function detrendPoly2(y) {
  const n = y.length;
  if (n < 3) return Float64Array.from(y);
  // Normalise the abscissa to [0,1] for conditioning.
  let s0 = n, s1 = 0, s2 = 0, s3 = 0, s4 = 0;
  let b0 = 0, b1 = 0, b2 = 0;
  for (let i = 0; i < n; i++) {
    const x = i / (n - 1);
    const x2 = x * x;
    s1 += x; s2 += x2; s3 += x2 * x; s4 += x2 * x2;
    b0 += y[i]; b1 += x * y[i]; b2 += x2 * y[i];
  }
  // Solve the 3x3 normal equations [s0 s1 s2; s1 s2 s3; s2 s3 s4] [a b c] = [b0 b1 b2].
  const m = [
    [s0, s1, s2, b0],
    [s1, s2, s3, b1],
    [s2, s3, s4, b2],
  ];
  // Gaussian elimination with partial pivoting.
  for (let col = 0; col < 3; col++) {
    let piv = col;
    for (let r = col + 1; r < 3; r++) if (Math.abs(m[r][col]) > Math.abs(m[piv][col])) piv = r;
    if (Math.abs(m[piv][col]) < 1e-12) continue;
    [m[col], m[piv]] = [m[piv], m[col]];
    for (let r = 0; r < 3; r++) {
      if (r === col) continue;
      const f = m[r][col] / m[col][col];
      for (let k = col; k < 4; k++) m[r][k] -= f * m[col][k];
    }
  }
  const a = m[0][3] / (m[0][0] || 1);
  const b = m[1][3] / (m[1][1] || 1);
  const c = m[2][3] / (m[2][2] || 1);
  const out = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const x = i / (n - 1);
    out[i] = y[i] - (a + b * x + c * x * x);
  }
  return out;
}

function hannWindow(n) {
  const w = new Float64Array(n);
  if (n < 2) { w.fill(1); return w; }
  for (let i = 0; i < n; i++) w[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (n - 1)));
  return w;
}

// |DFT|^2 of a windowed segment at frequency f.
function powerAt(seg, win, f, fs) {
  const w = (2 * Math.PI * f) / fs;
  let re = 0, im = 0;
  for (let n = 0; n < seg.length; n++) {
    const v = seg[n] * win[n];
    re += v * Math.cos(w * n);
    im -= v * Math.sin(w * n);
  }
  return re * re + im * im;
}

function median(arr) {
  if (!arr.length) return 0;
  const s = [...arr].sort((a, b) => a - b);
  const m = s.length >> 1;
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}

// Welch PSD across the candidate frequencies: average per-segment periodograms.
function welchSpectrum(sig, fs, freqs, segLen, hop) {
  const N = sig.length;
  const eff = Math.min(segLen, N);
  const win = hannWindow(eff);
  const starts = [];
  for (let s = 0; s + eff <= N; s += hop) starts.push(s);
  if (starts.length === 0) starts.push(0);

  const psd = new Float64Array(freqs.length);
  for (const s0 of starts) {
    const seg = sig.subarray(s0, s0 + eff);
    for (let k = 0; k < freqs.length; k++) psd[k] += powerAt(seg, win, freqs[k], fs);
  }
  for (let k = 0; k < psd.length; k++) psd[k] /= starts.length;
  return { psd, segments: starts.length, segLen: eff };
}

/**
 * Estimate respiration rate from RR-interval entries.
 * @param {{tMs:number, rr:number}[]} entries ascending by tMs (artifact-cleaned NN)
 * @param {object} [opts]
 * @returns {{ breathsPerMin:number|null, freqHz:number|null, confidence:number,
 *             snr:number|null, peakWidthHz:number|null, valid:boolean,
 *             preview:boolean, reason:string }}
 */
function estimateRespiration(entries, opts = {}) {
  const o = { ...DEFAULTS, ...opts };
  const fail = (reason) => ({
    breathsPerMin: null, freqHz: null, confidence: 0, snr: null,
    peakWidthHz: null, valid: false, preview: false, reason,
  });

  if (!Array.isArray(entries) || entries.length < o.minEntries) return fail('insufficient_intervals');
  const xs = entries.map((e) => e.tMs);
  const ys = entries.map((e) => e.rr);
  const span = xs[xs.length - 1] - xs[0];
  if (span < o.previewSpanMs) return fail('insufficient_span');

  const grid = resampleLinear(xs, ys, o.fs, xs[0], xs[xs.length - 1]);
  if (grid.length < 8) return fail('too_few_samples');
  const sig = detrendPoly2(grid);

  // Candidate frequencies over the search band.
  const freqs = [];
  for (let f = o.searchMin; f <= o.searchMax + 1e-9; f += o.step) freqs.push(f);

  const segLen = Math.round(o.segSec * o.fs);
  const hop = Math.max(1, Math.round(segLen * (1 - o.overlap)));
  const { psd } = welchSpectrum(sig, o.fs, freqs, segLen, hop);

  // Peak + robust noise floor (median is insensitive to the peak itself).
  let peakP = -1, peakK = -1;
  for (let k = 0; k < psd.length; k++) if (psd[k] > peakP) { peakP = psd[k]; peakK = k; }
  const floor = median(psd);
  const snr = floor > 0 ? peakP / floor : 0;
  const peakF = freqs[peakK];

  if (snr < o.minSnr) return fail('no_clear_peak');
  // Boundary guard: if the maximum sits on the first/last scanned bin the true
  // dominant oscillation is OUTSIDE the band (e.g. a strong sub-0.15 Hz Mayer
  // wave clipped to the lower edge), not a respiration peak. Reject it rather
  // than report the band edge with false confidence.
  if (peakK <= 0 || peakK >= psd.length - 1) return fail('peak_at_band_edge');

  // Peak sharpness: -3 dB (half-power) width around the peak. Narrow = a real
  // oscillation; broad = smeared noise.
  const half = peakP / 2;
  let lo = peakK, hi = peakK;
  while (lo > 0 && psd[lo] > half) lo--;
  while (hi < psd.length - 1 && psd[hi] > half) hi++;
  const peakWidthHz = (hi - lo) * o.step;

  // Confidence: SNR score (median-floor based) blended with sharpness.
  const snrScore = Math.max(0, Math.min(1, (snr - o.minSnr) / (12 - o.minSnr)));
  const widthScore = Math.max(0, Math.min(1, 1 - peakWidthHz / 0.08));
  let confidence = 0.7 * snrScore + 0.3 * widthScore;
  // De-weight peaks in the Mayer/LF zone unless they are very strong.
  if (peakF < o.mayerMax && snr < 6) confidence *= 0.4;

  const preview = span < o.minSpanMs;
  if (preview) confidence *= 0.6; // provisional: not enough breaths yet

  return {
    breathsPerMin: Math.round(peakF * 60 * 10) / 10,
    freqHz: Math.round(peakF * 1000) / 1000,
    confidence: Math.round(Math.max(0, Math.min(1, confidence)) * 100) / 100,
    snr: Math.round(snr * 10) / 10,
    peakWidthHz: Math.round(peakWidthHz * 1000) / 1000,
    valid: !preview,
    preview,
    reason: preview ? 'preview' : 'ok',
  };
}

module.exports = { estimateRespiration };

// --- self-test: synthesize an RR series modulated at a known rate ---
if (require.main === module) {
  function synth({ brPerMin = 14, seconds = 120, amp = 30, baseRr = 1000, noise = 5, driftAmp = 0, driftHz = 0.02 }) {
    const fHz = brPerMin / 60;
    const entries = [];
    let t = 0;
    while (t < seconds * 1000) {
      const resp = amp * Math.sin((2 * Math.PI * fHz * t) / 1000);
      const drift = driftAmp * Math.sin((2 * Math.PI * driftHz * t) / 1000);
      const rr = baseRr + resp + drift + (Math.random() - 0.5) * noise;
      t += rr;
      entries.push({ tMs: t, rr });
    }
    return entries;
  }
  const show = (label, est) =>
    console.log(
      `${label.padEnd(34)} br=${String(est.breathsPerMin).padStart(5)} conf=${String(est.confidence).padStart(4)} ` +
      `snr=${String(est.snr).padStart(5)} w=${est.peakWidthHz} valid=${est.valid} preview=${est.preview} (${est.reason})`,
    );
  console.log('# strong RSA');
  for (const target of [12, 15, 18]) show(`target ${target} br/min (amp40)`, estimateRespiration(synth({ brPerMin: target, amp: 40, noise: 5 })));
  console.log('# weak RSA (low RMSSD), noisy, with LF drift');
  show('amp6 noise2', estimateRespiration(synth({ brPerMin: 14, amp: 6, noise: 2, baseRr: 640 })));
  show('amp3 noise2', estimateRespiration(synth({ brPerMin: 14, amp: 3, noise: 2, baseRr: 640 })));
  show('amp3 noise2 +LFdrift15', estimateRespiration(synth({ brPerMin: 14, amp: 3, noise: 2, baseRr: 640, driftAmp: 15 })));
  console.log('# gating');
  show('30-60s window (preview)', estimateRespiration(synth({ brPerMin: 15, amp: 30, seconds: 45 })));
  show('short window (15s)', estimateRespiration(synth({ brPerMin: 15, amp: 30, seconds: 15 })));
}

File: app/android/app/src/main/java/dev/otake/rmssdh10n/hrv/Respiration.java
package dev.otake.rmssdh10n.hrv;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Respiration-rate estimation from RR intervals via RSA (Welch PSD). Originally a
 * port of src/respiration.js; the native engine is now Android's only compute
 * path, so this has DIVERGED from the JS reference on purpose to cut dropouts:
 *   - scan band extended down to 0.10 Hz so slow/relaxed breathing (6–9/min at
 *     rest, while coding, asleep) is no longer rejected as out-of-band. The
 *     0.10–0.15 Hz "slow" sub-band overlaps the Mayer wave (~0.1 Hz baroreflex),
 *     so it is admitted only under a stricter SNR + sharpness gate and reported
 *     at lower confidence (Task Force 1996 HF=0.15–0.40; Julien 2006 Mayer wave;
 *     HRV-biofeedback resonance ≈ 6/min = 0.1 Hz).
 *   - MIN_SNR lowered 2.5→2.0: a low-confidence number beats a blank (the UI
 *     shows signal quality separately). Natarajan 2021 gates RSA peaks by SNR.
 * The per-failure blanking is handled by the consumer (HrvEngine holds the last
 * good value and decays confidence). Pure DSP otherwise: linear resample to 4 Hz,
 * 2nd-order polynomial detrend, Hann-windowed Welch periodogram, peak vs
 * median-floor SNR + sharpness → signal-quality confidence.
 */
public final class Respiration {
    private static final double FS = 4;
    // SCAN_MIN is how low we look; SEARCH_MIN is the normal (HF) full-confidence
    // lower bound. [SCAN_MIN, SEARCH_MIN) is the gated slow-breathing sub-band.
    private static final double SCAN_MIN = 0.10, SEARCH_MIN = 0.15, SEARCH_MAX = 0.45, STEP = 0.005;
    private static final double SEG_SEC = 60, OVERLAP = 0.5;
    private static final double PREVIEW_SPAN_MS = 30000, MIN_SPAN_MS = 60000;
    private static final int MIN_ENTRIES = 20;
    private static final double MIN_SNR = 2.0;          // normal band
    private static final double SLOW_SNR = 4.0;         // slow band: stricter (Mayer guard)
    private static final double SLOW_MAX_WIDTH = 0.04;  // slow band: peak must be sharp

    public static final class Result {
        public final Double breathsPerMin; // null when no estimate
        public final Double confidence;     // signal quality 0..1
        public final boolean valid, preview;
        public final String reason;
        public final Double snr, freqHz, peakWidthHz; // diagnostics (null on fail)
        public final boolean slow;                     // peak fell in the slow sub-band
        Result(Double br, Double conf, boolean valid, boolean preview, String reason,
               Double snr, Double freqHz, Double peakWidthHz, boolean slow) {
            this.breathsPerMin = br; this.confidence = conf; this.valid = valid;
            this.preview = preview; this.reason = reason;
            this.snr = snr; this.freqHz = freqHz; this.peakWidthHz = peakWidthHz; this.slow = slow;
        }
    }

    private static Result fail(String reason) {
        return new Result(null, 0.0, false, false, reason, null, null, null, false);
    }

    /** entries ascending by tMs (artifact-cleaned NN beats). */
    public static Result estimate(double[] xs, double[] ys) {
        int n = xs.length;
        if (n < MIN_ENTRIES) return fail("insufficient_intervals");
        double span = xs[n - 1] - xs[0];
        if (span < PREVIEW_SPAN_MS) return fail("insufficient_span");

        double[] grid = resampleLinear(xs, ys, FS, xs[0], xs[n - 1]);
        if (grid.length < 8) return fail("too_few_samples");
        double[] sig = detrendPoly2(grid);

        List<Double> freqs = new ArrayList<>();
        for (double f = SCAN_MIN; f <= SEARCH_MAX + 1e-9; f += STEP) freqs.add(f);

        int segLen = (int) Math.round(SEG_SEC * FS);
        int hop = Math.max(1, (int) Math.round(segLen * (1 - OVERLAP)));
        double[] psd = welch(sig, FS, freqs, segLen, hop);

        double floor = median(psd);

        // Pick the strongest LOCAL maximum (not the global max) over the whole
        // band: a genuine oscillation is a bump (a bin above both neighbours)
        // whose fundamental dominates its harmonics, while the Mayer wave (~0.1 Hz
        // baroreflex) leaks only a monotonically-decaying tail into the low end —
        // which is NOT a local max and so is excluded by construction (it would
        // otherwise win as the global max and be reported as a bogus ~6/min
        // "breath", as seen live with snr≈52 at exactly 0.10 Hz).
        int peakK = bestLocalMax(psd, freqs, SCAN_MIN, SEARCH_MAX + 1e-9);
        if (peakK < 0) {
            double gf = freqs.get(argmax(psd));
            return new Result(null, 0.0, false, false, "no_clear_peak",
                    null, Math.round(gf * 1000) / 1000.0, null, false);
        }
        double snr = floor > 0 ? psd[peakK] / floor : 0;
        double peakF = freqs.get(peakK);
        double peakWidthHz = widthAt(psd, peakK);

        // Classify by frequency. The slow sub-band (0.10–0.15 Hz) overlaps the
        // Mayer wave, so it is admitted only under a strict SNR + sharpness gate
        // and reported at lower confidence; the HF band uses the normal gate.
        boolean slow = peakF < SEARCH_MIN;
        if (slow) {
            if (snr < SLOW_SNR || peakWidthHz > SLOW_MAX_WIDTH) {
                return new Result(null, 0.0, false, false, "slow_unconfirmed",
                        Math.round(snr * 10) / 10.0, Math.round(peakF * 1000) / 1000.0,
                        Math.round(peakWidthHz * 1000) / 1000.0, true);
            }
        } else if (snr < MIN_SNR) {
            return new Result(null, 0.0, false, false, "no_clear_peak",
                    Math.round(snr * 10) / 10.0, Math.round(peakF * 1000) / 1000.0,
                    Math.round(peakWidthHz * 1000) / 1000.0, false);
        }

        double snrScore = clamp01((snr - MIN_SNR) / (12 - MIN_SNR));
        double widthScore = clamp01(1 - peakWidthHz / 0.08);
        double confidence = 0.7 * snrScore + 0.3 * widthScore;
        if (slow) confidence *= 0.5; // honest: slow band is harder to separate from LF

        boolean preview = span < MIN_SPAN_MS;
        if (preview) confidence *= 0.6;

        double br = Math.round(peakF * 60 * 10) / 10.0;
        double conf = Math.round(clamp01(confidence) * 100) / 100.0;
        String reason = preview ? "preview" : (slow ? "ok_slow" : "ok");
        return new Result(br, conf, !preview, preview, reason,
                Math.round(snr * 10) / 10.0, Math.round(peakF * 1000) / 1000.0,
                Math.round(peakWidthHz * 1000) / 1000.0, slow);
    }

    private static double clamp01(double v) { return Math.max(0, Math.min(1, v)); }

    /** Strongest interior local maximum whose frequency is in [fLo, fHi), or -1.
     *  A local max (psd[k] above both neighbours) is a real oscillatory bump; a
     *  monotonic LF tail leaking into the band is excluded by construction. */
    private static int bestLocalMax(double[] psd, List<Double> freqs, double fLo, double fHi) {
        int best = -1; double bestP = -1;
        for (int k = 1; k < psd.length - 1; k++) {
            double f = freqs.get(k);
            if (f < fLo || f >= fHi) continue;
            if (psd[k] > psd[k - 1] && psd[k] >= psd[k + 1] && psd[k] > bestP) { bestP = psd[k]; best = k; }
        }
        return best;
    }

    /** -3 dB (half-power) width around bin k, in Hz. */
    private static double widthAt(double[] psd, int k) {
        double half = psd[k] / 2;
        int lo = k, hi = k;
        while (lo > 0 && psd[lo] > half) lo--;
        while (hi < psd.length - 1 && psd[hi] > half) hi++;
        return (hi - lo) * STEP;
    }

    private static int argmax(double[] a) {
        int idx = 0; double m = a.length > 0 ? a[0] : 0;
        for (int i = 1; i < a.length; i++) if (a[i] > m) { m = a[i]; idx = i; }
        return idx;
    }

    private static double[] resampleLinear(double[] xs, double[] ys, double fs, double t0, double tEnd) {
        double dtMs = 1000 / fs;
        int n = Math.max(2, (int) Math.floor((tEnd - t0) / dtMs) + 1);
        double[] out = new double[n];
        int j = 0;
        for (int i = 0; i < n; i++) {
            double t = t0 + i * dtMs;
            while (j < xs.length - 2 && xs[j + 1] < t) j++;
            double x0 = xs[j], x1 = xs[j + 1];
            double frac = x1 > x0 ? (t - x0) / (x1 - x0) : 0;
            out[i] = ys[j] + (ys[j + 1] - ys[j]) * Math.max(0, Math.min(1, frac));
        }
        return out;
    }

    private static double[] detrendPoly2(double[] y) {
        int n = y.length;
        if (n < 3) return Arrays.copyOf(y, n);
        double s0 = n, s1 = 0, s2 = 0, s3 = 0, s4 = 0, b0 = 0, b1 = 0, b2 = 0;
        for (int i = 0; i < n; i++) {
            double x = i / (double) (n - 1);
            double x2 = x * x;
            s1 += x; s2 += x2; s3 += x2 * x; s4 += x2 * x2;
            b0 += y[i]; b1 += x * y[i]; b2 += x2 * y[i];
        }
        double[][] m = {
            { s0, s1, s2, b0 },
            { s1, s2, s3, b1 },
            { s2, s3, s4, b2 },
        };
        for (int col = 0; col < 3; col++) {
            int piv = col;
            for (int r = col + 1; r < 3; r++) if (Math.abs(m[r][col]) > Math.abs(m[piv][col])) piv = r;
            if (Math.abs(m[piv][col]) < 1e-12) continue;
            double[] tmp = m[col]; m[col] = m[piv]; m[piv] = tmp;
            for (int r = 0; r < 3; r++) {
                if (r == col) continue;
                double f = m[r][col] / m[col][col];
                for (int k = col; k < 4; k++) m[r][k] -= f * m[col][k];
            }
        }
        double a = m[0][3] / (m[0][0] == 0 ? 1 : m[0][0]);
        double b = m[1][3] / (m[1][1] == 0 ? 1 : m[1][1]);
        double c = m[2][3] / (m[2][2] == 0 ? 1 : m[2][2]);
        double[] out = new double[n];
        for (int i = 0; i < n; i++) {
            double x = i / (double) (n - 1);
            out[i] = y[i] - (a + b * x + c * x * x);
        }
        return out;
    }

    private static double[] hann(int n) {
        double[] w = new double[n];
        if (n < 2) { Arrays.fill(w, 1); return w; }
        for (int i = 0; i < n; i++) w[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (n - 1)));
        return w;
    }

    private static double powerAt(double[] sig, int off, int len, double[] win, double f, double fs) {
        double w = (2 * Math.PI * f) / fs;
        double re = 0, im = 0;
        for (int nn = 0; nn < len; nn++) {
            double v = sig[off + nn] * win[nn];
            re += v * Math.cos(w * nn);
            im -= v * Math.sin(w * nn);
        }
        return re * re + im * im;
    }

    private static double[] welch(double[] sig, double fs, List<Double> freqs, int segLen, int hop) {
        int N = sig.length;
        int eff = Math.min(segLen, N);
        double[] win = hann(eff);
        List<Integer> starts = new ArrayList<>();
        for (int s = 0; s + eff <= N; s += hop) starts.add(s);
        if (starts.isEmpty()) starts.add(0);
        double[] psd = new double[freqs.size()];
        for (int s0 : starts) {
            for (int k = 0; k < freqs.size(); k++) psd[k] += powerAt(sig, s0, eff, win, freqs.get(k), fs);
        }
        for (int k = 0; k < psd.length; k++) psd[k] /= starts.size();
        return psd;
    }

    private static double median(double[] arr) {
        if (arr.length == 0) return 0;
        double[] s = arr.clone();
        Arrays.sort(s);
        int m = s.length >> 1;
        return (s.length % 2 == 1) ? s[m] : (s[m - 1] + s[m]) / 2.0;
    }

    private Respiration() {}
}
